Improve ImportExport with JavaBatch

prb112 opened the pull request:
- Change Runtime Check for ANTLR Version to 4.9.1
- Change confusing $import error message to reflect actual error
- Check the HttpStatusCode for $import/$export job creation for 201,
send useful message back
- Normalize the directory structure for the jbatch webapp
- Associate the jbatch webapp with the operation to normalize the code
- Move the IndexGenerator the the src/test/java to the tools package
- Remove prior FHIRBasicAuthenticator
- Tests Clients must use custom vm arguments to use the TLS key
    -Djavax.net.ssl.keyStore=/wlp/usr/servers/fhir-server/resources/security/fhirTrustStore.p12
    -Djavax.net.ssl.trustStore=/wlp/usr/servers/fhir-server/resources/security/fhirTrustStore.p12
    -Djavax.net.ssl.keyStoreType=p12
    -Djavax.net.ssl.trustStoreType=p12
    -Djavax.net.ssl.trustStorePassword=change-password
    -Djavax.net.ssl.keyStorePassword=change-password
- Remove the implied dependency on Apache CXF client in the bulkdata
client in test
- Remove DummpyImportExportImpl as it is unused
- Refactor the BulkDataClient to use Apache Http Client
- Created an Adapter to manage changes in the Configuration
- Added an Enabled Configuration Setting that is per tenant
- Added logging for the BulkData Client
- Refactor the batch-jobs definitions to include a consistent and
minimally sensitive Configuration.
- Normalize the OperationFields into a registry
- Update JobParameters to reflect the OperationFields
- Refactor Duplicate StorageType classes
- Move the tools to src/test/java
- Mark fhir-server provided in the webapp for bulkdata
- Bulk Data Audit Logging #1332
	- Add BulkAuditLogger and fhir-audit dependency
	- Add to Import
- NullPointer in ImportJobListener.afterJob #1361
	- Refactore the AfterJob ExitStatus and Reporter
	- Replicate and harden the partition handling in the
ImportPartitionMapper which feeds the transientUserData object
	- Occasionally we don't set the importResourceType on the
PartitionPlan.  I've added code to make sure this is set.
- Enforce StorageType source and destination for BulkData using
FHIR_BULKDATA_ALLOWED_TYPES #1982
	- Introduce configuration from System.env FHIR_BULKDATA_ALLOWED_TYPES
	- Add Operation specific checks for enabled/verify allowed source types
- Added Preflight Configuration Check for S3, Https, File
- Ambiguous dependencies when starting IBM FHIR Server 4.6.0 #1978
	- Handles the JobContext, StepContext to BulkDataContext using a
BatchContextAdapter
	- InjectionPoints are now specific versus generic String injection
points which cause the conflict.
	- Must inject the PartionPlan properties
- Remove the default Bucketname
- Refactoring of the fhirServer/bulkdata configuration #1390
	- Introduce Legacy Configuration
	- Introduce V2 Configuration
- Refactor Unique Calls to S3 Client into a Single Control Wrapper
- Exclude database-utils derby/db2 from the webapp.
- Job still in STARTED status after pod crash #1363
	- Add web.xml to block uncovered HTTP/S methods
	- Add ServletContextListener to Log Startup and hook into the local
BatchRuntime
- Added Logging Shunts where a try-catch are surrounding and rethrowing
the implemented JavaBatch Methods so the logging can push a message to
the main System.Out warning.
	- $import is now covered
	- Introduced StepChunkListener to unify the logging for Export/Import
- Increase logging in bulkdataclient
- Adding to TrustStore for S3
- Remove hardcoded password in the HttpWrapper
- Support bulk import/export from the local filesystem #1055
	- FileWrapper and a new file configuration enable writing to a local
file.
	- FilePreflight checks to see if the configuration supports write or
read
	- Also Add export to local file support to bulkdata export batchjob
#1211

Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

lmsurpre commented on operation/fhir-operation-bulkdata/pom.xml:
```diff
@@ -48,5 +48,13 @@
             <artifactId>jsonassert</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
```
not really related to this PR, but can we flip the `jakarta.ws.rs-api` above to provided?

tbieste commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/audit/BulkAuditLogger.java:
```diff
@@ -0,0 +1,312 @@
+     * - The end time of the request execution.
+     * @param responseStatus
+     * - The response status.
+     * @param location
```
Applies to several methods in this file:  Location and users params do not have hyphen in comments like other params.

prb112 replied:
Fixed, coming in PR

tbieste commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/export/group/resource/GroupHandler.java:
```diff
@@ -0,0 +1,150 @@
+import com.ibm.fhir.search.util.SearchUtil;
+
+/**
+ *
```
Please add class description.

prb112 replied:
next commit

tbieste commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/export/patient/resource/PatientResourceHandler.java:
```diff
@@ -0,0 +1,213 @@
+import com.ibm.fhir.search.util.SearchUtil;
+
+/**
+ *
```
Please add class description.

prb112 replied:
next commit

tbieste commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/export/writer/SparkParquetWriter.java:
```diff
@@ -4,7 +4,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
```
Update copyright.

prb112 replied:
in next commit

tbieste commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/jbatch/control/JobControllerBean.java:
```diff
@@ -0,0 +1,34 @@
+import javax.inject.Inject;
+
+/**
+ *
```
Please add class description.

prb112 replied:
I've since removed this class

tbieste commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/jbatch/export/data/CheckPointUserData.java:
```diff
@@ -4,7 +4,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
```
Update copyright.

prb112 replied:
next commit

tbieste commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/jbatch/export/data/TransientUserData.java:
```diff
@@ -4,7 +4,7 @@
  * SPDX-License-Identifier: Apache-2.0
```
Update copyright.

prb112 replied:
next commit

tbieste commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/jbatch/export/system/ChunkWriter.java:
```diff
@@ -0,0 +1,110 @@
+/*
+ * (C) Copyright IBM Corp. 2019, 2020
+ *
```
Update copyright.

prb112 replied:
Fixed in next commit

tbieste commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/jbatch/export/system/ExportJobListener.java:
```diff
@@ -0,0 +1,151 @@
+/*
+ * (C) Copyright IBM Corp. 2020
```
Update copyright.

prb112 replied:
Fixed in next commit

tbieste commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/jbatch/export/system/SystemExportPartitionMapper.java:
```diff
@@ -4,22 +4,26 @@
  * SPDX-License-Identifier: Apache-2.0
```
Update copyright.

prb112 replied:
Fixed in next commit

tbieste commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/jbatch/load/data/ImportCheckPointData.java:
```diff
@@ -4,7 +4,7 @@
  * SPDX-License-Identifier: Apache-2.0
```
Update copyright.

prb112 replied:
Fixed in next commit

tbieste commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/jbatch/load/data/ImportTransientUserData.java:
```diff
@@ -4,16 +4,19 @@
  * SPDX-License-Identifier: Apache-2.0
```
Update copyright.

prb112 replied:
Fixed in next commit

tbieste commented on fhir-bulkdata-webapp/src/test/java/com/ibm/fhir/bulkcommon/BulkDataUtilsTest.java:
```diff
@@ -14,15 +14,15 @@
 
```
Update copyright.

prb112 replied:
Fixed in next commit

tbieste commented on fhir-bulkdata-webapp/src/test/java/com/ibm/fhir/bulkcommon/SparkParquetWriterTest.java:
```diff
@@ -16,7 +16,7 @@
 import org.testng.annotations.Test;
 
-import com.ibm.fhir.jbatch.bulkdata.export.common.SparkParquetWriter;
+import com.ibm.fhir.bulkdata.export.writer.SparkParquetWriter;
```
Update copyright.

prb112 replied:
Fixed

tbieste commented on fhir-bulkdata-webapp/src/test/java/com/ibm/fhir/jbatch/bulkdata/tools/Main.java:
```diff
@@ -0,0 +1,339 @@
+/*
+ * (C) Copyright IBM Corp. 2020
```
Is this file needed, since it is commented out?  Otherwise, update copyright.

prb112 replied:
I'm going to remove it.  It was a prior test file.

tbieste commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/GroupExportOperation.java:
```diff
@@ -18,7 +18,7 @@
  * <code>$export</code>
```
Update copyright.

prb112 replied:
fixed

tbieste commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/PatientExportOperation.java:
```diff
@@ -18,7 +18,7 @@
  * <code>$export</code>
```
Update copyright.

prb112 replied:
fixed

tbieste commented on operation/fhir-operation-bulkdata/src/test/java/com/ibm/fhir/operation/bulkdata/model/BulkExportJobRestTest.java:
```diff
@@ -304,18 +304,11 @@ public void testBulkExportJobInstanceResponseParserAndGeneratorRoundTripEmpty()
     public void testBulkExportJobInstanceRequestWithSensitive() throws Exception {
         String jsonString =
```
Update copyright.

tbieste commented on operation/fhir-operation-bulkdata/src/test/java/com/ibm/fhir/operation/bulkdata/tool/helpers/dynamic/AgeRangeBloodPressureGroup.java:
```diff
@@ -4,7 +4,7 @@
  */
 
-package com.ibm.fhir.operation.bulkdata.helpers;
+package com.ibm.fhir.operation.bulkdata.tool.helpers.dynamic;
```
Update copyright.

prb112 replied:
I'll do all at the same time.  Thanks

tbieste commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/common/BulkDataUtils.java:
```diff
@@ -204,7 +153,7 @@ private static int getFhirResourceFromBufferReader(BufferedReader resReader, int
                     fhirResources.add(FHIRParser.parser(Format.JSON).parse(new StringReader(resLine)));
                     exported++;
-                    if (exported == Constants.IMPORT_NUMOFFHIRRESOURCES_PERREAD) {
+                    if (exported == adapter.getImportNumberOfFhirResourcesPerRead(null)) {
```
Maybe change to ">=" just in case somewhere/somehow getImportNumberOfFhirResourcesPerRead
 returned a different value and we loop forever.

prb112 replied:
it is a constant underneath.  However, it could change based on the interface.  Many thanks for the catch.

prb112 replied:
Fixed

tbieste reviewed (commented):
Left minor comments primarily around copyright year updates, etc.

lmsurpre commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/audit/BulkAuditLogger.java:
```diff
@@ -0,0 +1,311 @@
+     * @return AuditLogEntry an audit log entry.
+     */
+    private AuditLogEntry createAuditLogEntry(AuditLogEventType eventType, Resource resource, Date startTime, Date endTime, Response.Status responseStatus,
+        String location, String users) {
```
Applies to a number of long method declarations:  please use double-indent to offset from the first expression of the method

lmsurpre commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/audit/BulkAuditLogger.java:
```diff
@@ -0,0 +1,311 @@
+     * @throws Exception
+     */
+    public void logUpdateOnImport(Resource oldResource, Resource updatedResource, Date startTime, Date endTime, Response.Status responseStatus, String location,
+        String users)
+        throws Exception {
```
Applies to a number of long method declarations:  please use double-indent to offset from the first expression of the method

lmsurpre commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/common/BulkDataUtils.java:
```diff
@@ -59,30 +45,33 @@
 import com.ibm.fhir.model.resource.Resource;
 import com.ibm.fhir.model.util.FHIRUtil;
 import com.ibm.fhir.model.util.ModelSupport;
+import com.ibm.fhir.operation.bulkdata.config.ConfigurationAdapter;
+import com.ibm.fhir.operation.bulkdata.config.ConfigurationFactory;
```
I've always wanted some common spot for the stuff thats shared between fhir-operation-bulkdata and fhir-bulkdata-webapp; maybe another day :-)

prb112 replied:
This BulkDataUtils is really focused on S3 operations. 
I pushed the helper methods in the corresponding provider classes (StorageProvider)

lmsurpre commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/jbatch/export/checkpoint/CheckPointAlgorithm.java:
```diff
@@ -0,0 +1,73 @@
+ * BulkData Export Custom CheckpointAlgorithm which considers COS size requirements while checkpointing.
+ */
+@Dependent
+public class CheckPointAlgorithm implements CheckpointAlgorithm {
```
having the instance class name vary from the interface name only by casing is a little odd.  is there any distinguishing feature from other checkpoint algorithm which we can use in the name?  maybe `BulkExportCheckpointAlgorithm`?

prb112 replied:
yes - totally agree.  I'll make that change

lmsurpre commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/jbatch/export/data/CheckPointUserData.java:
```diff
@@ -1,18 +1,17 @@
- *
+ * Bulk Export Job Checkpoint data.
  */
 public class CheckPointUserData implements java.io.Serializable {
```
unrelated to your change, but wondering if we should follow the framework's precedence here and use `CheckpointUserData` rather than `CheckPointUserData`

prb112 replied:
I have no preferences - which one ? you pick ;-)

lmsurpre replied:
I vote to update any classnames that have `CheckPoint` and change them to `Checkpoint`

lmsurpre commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/jbatch/export/fast/NopWriter.java:
```diff
@@ -27,7 +27,7 @@
     JobContext jobContext;
 
-    /**
+    /*
```
why not javadoc?

prb112 replied:
it's a private method

prb112 replied:
the next  method was private

lmsurpre replied:
private methods can have javadoc.  it makes it come up nice from your IDE.  it doesn't go into the published javadoc.  I'd vote to keep using javadoc-style comments for private methods.

lmsurpre commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/FileProvider.java:
```diff
@@ -0,0 +1,153 @@
+        Path p1 = Paths.get(fn);
+        try {
+            // This is a trap. Be sure to mark CREATE and APPEND.
+            // We do create empty files.
```
under what conditions do we create empty files?

prb112 replied:
0 resourced files. 
E.g. export all resources and we have zero for parameters. 

I decided too put it in writeResources.

prb112 replied:
I just changed this slightly so we don't create empty files.

lmsurpre replied:
sounds good

lmsurpre commented on fhir-path/src/main/java/com/ibm/fhir/path/FHIRPathLexer.java:
```diff
@@ -1,13 +1,17 @@
 import org.antlr.v4.runtime.dfa.DFA;
-import org.antlr.v4.runtime.misc.*;
+
+//Manually changed to 4.9.1
```
seems a bit odd to have these fhirpath antlr changes mixed in with the other ones.  I was thinking John already addressed this, but I guess that was just for the model?

prb112 replied:
oh - good find.   yeah - I  did it prior to John, it turns out there really aren't any changes with 4.9.1

lmsurpre commented on fhir-server-test/src/test/resources/test.properties:
```diff
@@ -39,11 +39,12 @@ fhirclient.http.return.pref = minimal
 test.kafka.enabled = false
-test.bulkdata.import.enabled = false
-test.bulkdata.export.enabled = false
-test.bulkdata.useminio = false
-test.bulkdata.minio.username = minio
-test.bulkdata.minio.password = change-password
```
do we still exercise minio during the CI, or is it only the file-based storage provider that gets tested now?

prb112 replied:
It's  only the file-base provider. i can  enable minio too  - I'd just have to include a different set of tests. 
We could do that with a separate issue.  Let me  know and I can create it.

lmsurpre replied:
I think it would be nice to exercise both providers (since we already did the work to enable minio).  Agree it can be a future thing.

prb112 replied:
both providers are in the IT tests

lmsurpre commented:
question:  what is the change to `fhir-server/liberty-config/resources/security/fhirTrustStore.p12` ?

lmsurpre commented on fhir-server/liberty-config/server.xml:
```diff
@@ -14,7 +14,7 @@
              but we include it here to avoid NoClassDefFound in our classes that *can* use it -->
         <feature>mpJwt-1.1</feature>
-    </featureManager>
+	</featureManager>
```
looks like a tab snuck in.  I'm assuming its accidental:
```suggestion
    </featureManager>
```

lmsurpre commented on operation/fhir-operation-bulkdata/src/docs/Configuration.md:
```diff
@@ -0,0 +1,79 @@
+The following is a brief template of the V2 Configuration. 
```
I think its the first time I've seen us put info under `src/docs`.  Do we want to adopt this pattern going forward?
If not, maybe just stuff it into the fhir-operation-bulkdata README?

prb112 replied:
moved/changed

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/BulkDataResourceProvider.java:
```diff
@@ -1,16 +0,0 @@
-
-import com.ibm.fhir.registry.util.PackageRegistryResourceProvider;
-
-public class BulkDataResourceProvider extends PackageRegistryResourceProvider {
```
I believe this registry resource provider is what we were using to advertise our support of the bulkdata spec in our CapabilityStatement.  Are you still advertising that capability, but in a new way?
See https://github.com/IBM/FHIR/issues/1315 for more info.

lmsurpre replied:
ah, now I see it was just moved to `ResourceProvider` under subpackage `provider`.

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/client/BulkDataClient.java:
```diff
@@ -652,53 +771,81 @@ public String submitImport(String inputFormat, String inputSource, List<Input> i
     }
 
-    /**
+    /*
```
why not javadoc?

prb112 replied:
it really isn't intended  too be publicaly readable.  It's also a private method.  thus the `/*`

lmsurpre replied:
same comment as the other one...I think its totally acceptable to use javadoc-style comments on private methods

lmsurpre replied:
```suggestion
    /**
```

prb112 commented:
> question: what is the change to `fhir-server/liberty-config/resources/security/fhirTrustStore.p12` ?

I added the key's signer to the truststore. (it was an annoying thing with httpclient) 
I think I also added the ibm  cloud db ca

lmsurpre commented on operation/fhir-operation-bulkdata/src/test/sh/export-group-1.sh:
```diff
@@ -0,0 +1,30 @@
+    --header 'Content-Type: application/fhir+json' -k \
+    -u "fhiruser:${PASS}" -v --header "X-FHIR-TENANT-ID: ${TENANT_ID}"
+
+# 4 - Check the file that is output.
```
Is there a way to automate the checking of the file that is output? (for future consideration)

prb112 replied:
It really depends on the storage.

prb112 replied:
I gave an example in this one.

lmsurpre commented on fhir-bulkdata-webapp/pom.xml:
```diff
@@ -45,6 +45,23 @@
             <groupId>org.glassfish</groupId>
             <artifactId>jakarta.json</artifactId>
         </dependency>
+        <dependency>
+            <!-- This is provided, yet never used. -->
+            <groupId>${project.groupId}</groupId>
+            <artifactId>fhir-server</artifactId>
+            <version>${project.version}</version>
+            <scope>provided</scope>
+        </dependency>
```
wondering what this is for... are we just using it for its transitive dependency on other things marked provided?

prb112 replied:
since we pick up operation-bulkdata, I don't want to pull in fhir-server.  I marked as provided, and it is not used.

lmsurpre commented on fhir-server/liberty-config/config/default/fhir-server-config-db2.json:
```diff
@@ -117,30 +117,83 @@
+                "default" : {
+                    "type": "file",
+                    "_type": "ibm-cos|aws-s3|file|https",
+                    "valid-base-urls": [],
```
previously we've stuck to CamelCase for config props, but I think I'm OK with consistently using kabob-case on these culk properties...anyone disagree?

prb112 replied:
flipping to camelcase is quick  change.   just let me know

lmsurpre commented on fhir-server/liberty-config/config/default/fhir-server-config-db2.json:
```diff
@@ -117,30 +117,83 @@
+                    "password": "change-password",
+                    "truststore": "resources/security/fhirTrustStore.p12",
+                    "truststore-password": "change-password", 
+                    "trust-all": true
```
what does this do?

prb112 replied:
sets the client ssl to trust the ca and hostname in the http client used in httpwrapper

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/config/preflight/impl/FilePreflight.java:
```diff
@@ -0,0 +1,82 @@
+                throw util.buildExceptionWithIssue("The location for the bulkdata operation is not found", IssueType.NOT_FOUND);
+            }
+        } else {
+            throw util.buildExceptionWithIssue("No File Base Configured for FHIR bulkdata operation", IssueType.INVALID);
```
INVALID is used to indicate that the request was invalid, whereas I think this is a server-side configuration error.  We should be sure that these map to HTTP 500 errors instead, such as by using 
```suggestion
            throw util.buildExceptionWithIssue("No File Base Configured for FHIR bulkdata operation", IssueType.EXCEPTION);
```

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/config/preflight/impl/FilePreflight.java:
```diff
@@ -0,0 +1,82 @@
+                }
+            }
+            if (!accessible) {
+                throw util.buildExceptionWithIssue("The location for the bulkdata operation is not found", IssueType.NOT_FOUND);
```
NOT_FOUND is typically mapped into a HTTP 404 error.  In this case, its actually the backend file storageProvider that is misconfigured and so I think we should make sure this maps to an HTTP 500 instead.
```suggestion
                throw util.buildExceptionWithIssue("The location for the bulkdata operation is not found", IssueType.NO_STORE);
```

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/config/preflight/impl/FilePreflight.java:
```diff
@@ -0,0 +1,82 @@
+    public void checkStorageAllowed(StorageDetail storageDetail) throws FHIROperationException {
+        if (storageDetail != null && !StorageType.FILE.value().equals(storageDetail.getType())){
+            CommonUtil util = new CommonUtil();
+            throw util.buildExceptionWithIssue("File: Configuration not set to import from storageDetail '" + getSource() + "'", IssueType.INVALID);
```
INVALID maps to 400 Bad Request.  Is this scenario a bad request or a misconfigured server?

prb112 replied:
it's actually both.  It means that  the request did not match the server settings. and that may mean the server has a bad configuration

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/config/preflight/impl/S3Preflight.java:
```diff
@@ -0,0 +1,159 @@
+            List<Future<Boolean>> futures = executor.invokeAll(callables, 60, TimeUnit.SECONDS);
+            for (Future<Boolean> future : futures) {
+                if (!future.get()) {
+                    throw export.buildOperationException("Unable to access s3 source or outcome during timeout", IssueType.INVALID);
```
```suggestion
                    throw export.buildOperationException("Unable to access s3 source or outcome during timeout", IssueType.EXCEPTION);
```

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/config/preflight/impl/S3Preflight.java:
```diff
@@ -0,0 +1,159 @@
+                }
+            }
+        } catch (ExecutionException ee) {
+            throw export.buildOperationException("Failed to execute the s3 access, check the s3 configuration", IssueType.INVALID);
```
```suggestion
            throw export.buildOperationException("Failed to execute the s3 access, check the s3 configuration", IssueType.NO_STORE);
```

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/config/preflight/impl/S3Preflight.java:
```diff
@@ -0,0 +1,159 @@
+        } catch (ExecutionException ee) {
+            throw export.buildOperationException("Failed to execute the s3 access, check the s3 configuration", IssueType.INVALID);
+        } catch (InterruptedException e) {
+            throw export.buildOperationException("Timeout hit trying to access s3, check the s3 configuration", IssueType.INVALID);
```
```suggestion
            throw export.buildOperationException("Timeout hit trying to access s3, check the s3 configuration", IssueType.NO_STORE);
```

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/config/preflight/impl/S3Preflight.java:
```diff
@@ -0,0 +1,159 @@
+            String apiKey = adapter.getStorageProviderAuthTypeIamApiKey(source);
+            String resourceId = adapter.getStorageProviderAuthTypeIamApiResourceInstanceId(source);
+            if (apiKey == null || resourceId == null || apiKey.isEmpty() || resourceId.isEmpty()) {
+                throw export.buildOperationException("bad configuration for the iam configuration", IssueType.INVALID);
```
```suggestion
                throw export.buildOperationException("bad configuration for the iam configuration", IssueType.EXCEPTION);
```

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/config/preflight/impl/S3Preflight.java:
```diff
@@ -0,0 +1,159 @@
+            String accessKey = adapter.getStorageProviderAuthTypeHmacAccessKey(source);
+            String secretKey = adapter.getStorageProviderAuthTypeHmacSecretKey(source);
+            if (accessKey == null || secretKey == null || accessKey.isEmpty() || secretKey.isEmpty()) {
+                throw export.buildOperationException("bad configuration for the hmac configuration", IssueType.INVALID);
```
```suggestion
                throw export.buildOperationException("bad configuration for the hmac configuration", IssueType.EXCEPTION);
```

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/config/preflight/impl/S3Preflight.java:
```diff
@@ -0,0 +1,159 @@
+            String user = adapter.getStorageProviderAuthTypeUsername(source);
+            String password = adapter.getStorageProviderAuthTypePassword(source);
+            if (user == null || password == null || user.isEmpty() || password.isEmpty()) {
+                throw export.buildOperationException("bad configuration for the basic configuration", IssueType.INVALID);
```
```suggestion
                throw export.buildOperationException("bad configuration for the basic configuration", IssueType.EXCEPTION);
```

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/config/preflight/impl/S3Preflight.java:
```diff
@@ -0,0 +1,159 @@
+                throw export.buildOperationException("bad configuration for the basic configuration", IssueType.INVALID);
+            }
+        } else {
+            throw export.buildOperationException("Failed to specify the source or outcome bucket's authentication mechanism", IssueType.INVALID);
```
```suggestion
            throw export.buildOperationException("Failed to specify the source or outcome bucket's authentication mechanism", IssueType.EXCEPTION);
```

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/config/preflight/impl/S3Preflight.java:
```diff
@@ -0,0 +1,159 @@
+            ConfigurationAdapter adapter = ConfigurationFactory.getInstance();
+            url = adapter.getStorageProviderEndpointExternal(source);
+            if (url == null || url.isEmpty()) {
+                throw export.buildOperationException("endpoint-internal is undefined.", IssueType.INVALID);
```
```suggestion
                throw export.buildOperationException("endpoint-internal is undefined.", IssueType.EXCEPTION);
```

lmsurpre commented on fhir-bulkdata-webapp/src/main/resources/META-INF/batch-jobs/FhirBulkExportFastJob.xml:
```diff
@@ -0,0 +1,59 @@
+                <properties>
+                </properties>
+            </writer>
+            <checkpoint-algorithm ref="com.ibm.fhir.bulkdata.jbatch.export.fast.checkpoint.ResourceExportCheckpointAlgorithm">
```
The class ref doesn't match its updated package name.
```suggestion
            <checkpoint-algorithm ref="com.ibm.fhir.bulkdata.export.fast.checkpoint.ResourceExportCheckpointAlgorithm">
```

prb112 replied:
merged locally

lmsurpre commented:
One thing I'm not sure about is requiring the client to pass us `X-FHIR-BULKDATA-PROVIDER` to select between the various configurations.  Is it properly tenant-enanbled now so we can at least have a different `default` for each tenant?

prb112 commented:
> One thing I'm not sure about is requiring the client to pass us `X-FHIR-BULKDATA-PROVIDER` to select between the various configurations.
> Because import and export storageProviders aren't distinct, won't this prevent us from using a different default storageProvider for import than we use for export?
> Is it properly tenant-enanbled now so we can at least have a different `default` for each tenant?

by default it is fully tenant aware.  Yes.  the import and  export share the same default.  This is intentional.  If a client wants to split them out they can specify.  Else, we could provide TWO configurations which  are duplicate for the same purposes.  I think this is the best option.  The second best option I  considered was using "export" and "import" for  the default names, once  again,  we duplicate the configuration.  

However, we could override the defaults for the system.  (slight addition where we point to the default in the configuration. 

Let me know, I can easily add it.

lmsurpre commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/jbatch/context/BatchContextAdapter.java:
```diff
@@ -0,0 +1,189 @@
+        context(ctx);
+        source(ctx);
+        audit(ctx);
+        search(ctx);
```
I don't know if we need to set the search-related parameters or not (it doesn't support _typefilter), but we *definitely* need to set the systemWrite ones so that we get our cosPathPrefix.
```suggestion
        systemWrite(ctx);
        search(ctx);
```

lmsurpre replied:
I'm also confused about why cosPathPrefix is associated with system export...shouldn't we be setting that for *all* exports?  Did you do something different to set it in the Group and Patient exports?

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/config/impl/AbstractSystemConfigurationImpl.java:
```diff
@@ -0,0 +1,283 @@
+        return coreCosMaxSize;
+    }
+
+    private static final int defaultCoreCosMaxSize() {
+        final String PATH = "fhirServer/bulkdata/core/cos/max-size";
+        return FHIRConfigHelper.getIntProperty(PATH, DEFAULT_COSFILE_MAX_SIZE);
+    }
```
I'm thinking that we should probably make this a `long`.
The default fits safely into an int, but--if we keep this in bytes--then its not too far off from Integer.MAX_VALUE:
`209715200`  (our default; 200MiB)
`2147483647` (Integer.MAX_VALUE; 2048 MiB)

Unfortunately, our fhir-config stuff doesn't currently support Long properties, so we'd either need to add that support or decide to represent the value in KB or MB (or MiB) in our config.

lmsurpre replied:
I think my vote would be to just specify it in MiB in our config.  I don't think a max-size of <1 MiB makes any sense, and I don't think there's much value in allowing more granular values.

prb112 replied:
I've changed to MIB in config

prb112 replied:
It's in the next commit

lmsurpre commented:
> However, we could override the defaults for the system. (slight addition where we point to the default in the configuration.
> Let me know, I can easily add it.

I opened https://github.com/IBM/FHIR/issues/2036 for it.

lmsurpre commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/export/system/resource/SystemExportResourceHandler.java:
```diff
@@ -0,0 +1,90 @@
+                throw e;
+            }
+        }
+        chunkData.setCurrentUploadResourceNum(chunkData.getCurrentUploadResourceNum() + resSubTotal);
+        chunkData.addCurrentUploadSize(chunkData.getBufferStream().size());
+        chunkData.setTotalResourcesNum(chunkData.getTotalResourcesNum() + resSubTotal);
```
could use the addX flavor for all three (for consistency):
```suggestion
        chunkData.addCurrentUploadResourceNum(resSubTotal);
        chunkData.addCurrentUploadSize(chunkData.getBufferStream().size());
        chunkData.addTotalResourcesNum(resSubTotal);
```

lmsurpre replied:
could use the addX flavor for all three (for consistency):
```suggestion
        chunkData.addCurrentUploadResourceNum(resSubTotal);
        chunkData.addCurrentUploadSize(chunkData.getBufferStream().size());
        chunkData.addTotalResourcesNum(resSubTotal);
```

prb112 commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/jbatch/export/fast/ResourcePayloadReader.java:
```diff
@@ -491,9 +384,12 @@ public Boolean processPayload(ResourcePayload t) {
+                //
+                // @implNote outputStream.size() is not set for the current resource yet set at this point of the current Resource, comparing it
+                //  can lead to unintended results with double counting.
+                if (this.uploadId != null && (currentObjectSize > maxObjectSize
```
Here is the point of discussion

prb112 replied:
In prior it was comparing against an already populated outputStream.size()
note the outputstream is not accumulated until line #405

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/config/impl/AbstractSystemConfigurationImpl.java:
```diff
@@ -152,7 +152,7 @@ public long getCoreCosMaxResources() {
     private static final long defaultCoreCosMaxResources() {
         final String PATH = "fhirServer/bulkdata/core/cos/maxResourcesThreshold";
-        return 1000 * 1000 * FHIRConfigHelper.getIntProperty(PATH, DEFAULT_COSFILE_MAX_RESOURCESNUMBER);
+        return 1000 * FHIRConfigHelper.getIntProperty(PATH, DEFAULT_COSFILE_MAX_RESOURCESNUMBER);
```
do we need to use `1000L` to force it to autobox to a long?  or just to make it more clear?

prb112 replied:
sure

lmsurpre commented on fhir-server/liberty-config/config/default/fhir-server-config-db2.json:
```diff
@@ -117,30 +117,81 @@
+                },
+                "minio" : {
+                    "type": "aws-s3",
+                    "bucket-name": "fhirbulkdata",
```
```suggestion
                    "bucketName": "fhirbulkdata",
```

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/config/impl/V2ConfigurationImpl.java:
```diff
@@ -0,0 +1,178 @@
+
+    @Override
+    public String getStorageProviderBucketName(String provider) {
+        return FHIRConfigHelper.getStringProperty("fhirServer/bulkdata/storageProviders/" + provider + "/bucketBame", null);
```
```suggestion
        return FHIRConfigHelper.getStringProperty("fhirServer/bulkdata/storageProviders/" + provider + "/bucketName", null);
```

lmsurpre commented on operation/fhir-operation-bulkdata/src/test/resources/config/default/fhir-server-config.json:
```diff
@@ -0,0 +1,42 @@
+            "storageProviders": {
+                "default" : {
+                    "type": "ibm-cos",
+                    "validBaseRrls": [],
```
```suggestion
                    "validBaseUrls": [],
```

lmsurpre commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java:
```diff
@@ -0,0 +1,441 @@
+                break;
+            case FHIRMediaType.APPLICATION_NDJSON:
+            default:
+                if (chunkData.getPageNum() > chunkData.getLastPageNum() || chunkData.isFinishCurrentUpload()) {
```
Looks like we accidentally lost our "do we have enough bytes to start a partUpload" check from here.
I believe what we wanted to do is remove the `chunkData.getPageNum() > chunkData.getLastPageNum()` check from here, but *leave* the `chunkData.getBufferStream().size() > COS_PART_MINIMALSIZE` one
(see https://github.com/IBM/FHIR/pull/1999/commits/61e8e40f7944aff4d8083f7e89e7b2c23b54e2e2#diff-83b573e778b2ec67186ea08aa4515c6d7e6b3583c592abf3fc44cbb3240a7f06L343).
